Peer db new serialization

paninaro opened the pull request:
Trello task: https://trello.com/c/bstGWYhf/1381-replace-peer-dbs-with-flat-file

Switches to a new serialization format for the peer data that was previously saved to a sqlite db. The peer data that was previously saved to `peer_db_table.sqlite` and `wallet_peers.sqlite` will now be stored in `peers.dat` and `wallet_peers.dat` respectively.

**Note**: Existing peer data will be migrated to the new format. Existing peer sqlite DBs are left untouched in case users need to downgrade.

AddressManagerStore has been simplified by removing the aiosqlite logic. Deserialization is performed synchronously, while serialization is asynchronous.

`initial-config.yaml` has been updated such that `peer_db_path` and `wallet_peers_path` have been replaced by `peers_file_path` and `wallet_peers_file_path`. While the initial config has been updated, no attempt is made to update existing config.yaml files. By leaving existing config.yaml files as-is, we can support downgrading.

The PeerStoreResolver class has been added to determine the appropriate peer file path given data from the config. The resolved file path is stored in the config object that is passed in during initialization. Currently, this resolved value is not written back to config.yaml, though this is possible if we want to provide a migration option.

Tests have been added to cover PeerStoreResolver and the updates to AddressManagerStore.

paninaro committed:
Serialize/deserialize peer data alongside existing sqlite implementation (to be removed)

paninaro committed:
Simplified AddressManagerStore. No longer uses sqlite and is no longer async.

paninaro committed:
Removed aiosqlite usage from AddressManagerStore.
Added PeerStoreResolver class to determine the appropriate location for "peers.dat"
Updated initial-config.yaml to include "peers_file_path" default, replacing "peer_db_path" (similar change for "wallet_peers_path")

paninaro committed:
Minor comment changes/additions

paninaro committed:
Added migration from sqlite peer db.
Made AddressManagerStore's serialization async as it was right at the edge of blocking for too long.

arvidn commented on chia/server/node_discovery.py:
```diff
@@ -80,6 +82,20 @@ def __init__(
+            if address_manager is not None:
+                self.log.info(f"Writing migrated peer data to {self.peers_file_path}")
+                # Write the AddressManager data to the new peers file
+                await AddressManagerStore.serialize(address_manager, self.peers_file_path)
```
if we get here without an error, shouldn't `self.legacy_peer_db_path` be set to `None`? I suppose this is only called on `start()` so maybe not. However, we don't delete (or move aside) the old DB. How do we record that we've already migrated, to avoid doing it again next time we restart (and clobber our new peer list with an old one)?

paninaro replied:
I'll add a `migrated` instance variable to track this within the same run to make things a little tidier. Though, once migrated, the new file exists, so we won't reattempt migration.

arvidn replied:
I see, we should probably check for its existence first then (before the db)

arvidn commented on chia/server/address_manager_sqlite_store.py:
```diff
@@ -0,0 +1,119 @@
+    @classmethod
+    def _create_address_manager(
+        cls, metadata: Dict[str, str], nodes: List[Node], new_table_entries: List[Table]
+    ) -> AddressManager:
```
Doesn't the `AddressManager` have a constructor that lets it be constructed from these dicts and lists?
It seems suspicious that you duplicated the code below, rather than moved it. If it's still in its original location, is there a reason you can't use that original copy of it?

paninaro replied:
I wanted the migration code to be logically identical to what it was before. The reason for duplicating and not reusing/moving is so that changes can be made in AddressManagerStore going forward without affecting migration.

arvidn reviewed (commented):
I think it's important that you add tests for these new features

arvidn commented on chia/util/write_file.py:
```diff
@@ -0,0 +1,24 @@
+from typing import Union
+
+
+def write_file(file_path: Path, data: Union[str, bytes], *, file_mode: int = 0o600):
```
aren't there named constants for `0o600`?

I also think there should be unit tests for this function

paninaro replied:
Tests added.

Re: The named constants for the permission mask. I strongly prefer the octal representation, which I think is fairly universally understood by devs. The alternative is to explode out a bitmask of similarly named constants like `stat.S_IRUSR | stat.S_IWUSR |  stat.S_IRGRP | stat.S_IWGRP` etc. I find it far more readable to use the octal representation, and less prone to accidentally choose the wrong permission bit(s). The octal representation is interoperable with the named constants.

arvidn commented on chia/util/write_file.py:
```diff
@@ -0,0 +1,24 @@
+    os.makedirs(os.path.dirname(file_path), mode=dir_perms, exist_ok=True)
+
+    temp_path: Path = file_path.with_suffix("." + str(os.getpid()))
+    mode: str = "w" if type(data) == str else "wb"
```
I would have expected you would need `w+` and `wb+` respectively. what if a previous error left the file (with the pid suffix) around? it would need to be truncated.

There should probably be a test for that.

paninaro replied:
I've switched it to use w+ and w+b. Existing file issue has been worked-around by using the tempfile functionality to guarantee a uniquely named temp file.

arvidn commented on chia/util/write_file.py:
```diff
@@ -0,0 +1,24 @@
+
+    temp_path: Path = file_path.with_suffix("." + str(os.getpid()))
+    mode: str = "w" if type(data) == str else "wb"
+    with open(os.open(str(temp_path), os.O_CREAT | os.O_EXCL | os.O_WRONLY, file_mode), mode) as f:
```
I see, you're using `O_EXCL` here, which would throw an exception if the file already existed, right?
Do we just give up in that case and fail everything?

paninaro replied:
Changed to use the tempfile module to guarantee a unique temp file for writing.

arvidn commented on chia/util/write_file.py:
```diff
@@ -0,0 +1,24 @@
+    try:
+        os.replace(str(temp_path), file_path)
+    except PermissionError:
+        shutil.move(str(temp_path), str(file_path))
```
if this call fails, we don't delete the temporary file. should we?

arvidn replied:
I'm pretty sure that both this and `os.replace()` will fail on windows if some other process (or our process) has the file open. Perhaps it would be reasonable to retry a few times, if you make `write_file()` async, you could async sleep for 0.5 seconds or so and retry a few times, before failing.

paninaro replied:
There were some edge cases that caused issues on Windows, as you hinted at. I've added retry logic, and updated the code such that the temp file is removed if the move fails.

Several tests to handle os.replace failures. The tests pass for me on macOS and Windows.

arvidn commented on chia/util/write_file.py:
```diff
@@ -0,0 +1,24 @@
+    temp_path: Path = file_path.with_suffix("." + str(os.getpid()))
+    mode: str = "w" if type(data) == str else "wb"
+    with open(os.open(str(temp_path), os.O_CREAT | os.O_EXCL | os.O_WRONLY, file_mode), mode) as f:
+        f.write(data)
```
how do we feel about using `aiofiles` and perform this write asynchronously instead?

i.e.
```
await f.write(data)
```

paninaro replied:
I had considered aiofiles, but initially wanted to keep the change as small/tidy as possible. That's changed now, so I've pulled in aiofiles.

arvidn commented on chia/server/address_manager_store.py:
```diff
@@ -10,10 +8,54 @@
+
+async def makePeerDataSerialization(
+    metadata: List[Tuple[str, Any]], nodes: List[Tuple[int, ExtendedPeerInfo]], new_table: List[Tuple[int, int]]
+) -> PeerDataSerialization:
```
is there a reason not to just return `bytes` here directly? as far as I can tell, in all places we call this function we immediately turn the `PeerDataSerialization` object into `bytes`

paninaro replied:
Sounds reasonable

arvidn commented on chia/server/address_manager_store.py:
```diff
@@ -145,59 +130,104 @@ async def serialize(self, address_manager: AddressManager):
+        Read the peers file and return the data as a PeerDataSerialization object.
+        """
+        with open(peers_file_path, "rb") as f:
+            return PeerDataSerialization.from_bytes(f.read())
```
I would also imagine this would by `async` and use `await f.read()` in an ideal world. Was that considered?

paninaro replied:
Updated to use aiofiles.

arvidn reviewed (commented):
did you profile this patch? if so, did you measure an improvement?

paninaro commented on chia/server/node_discovery.py:
```diff
@@ -80,6 +82,20 @@ def __init__(
 
+    async def migrate_address_manager_if_necessary(self) -> None:
+        if self.legacy_peer_db_path is None or not self.legacy_peer_db_path.exists() or self.peers_file_path.exists():
+            return
```
I'd be more comfortable with adding a new instance variable like `self.migrated = True` to allow us to exit early if this method is called more than once. I'll make that change.

arvidn replied:
Makes sense. Is there a mechanism to prevent migrating on every startup though?

arvidn commented on chia/server/node_discovery.py:
```diff
@@ -80,6 +82,20 @@ def __init__(
+                # Write the AddressManager data to the new peers file
+                await AddressManagerStore.serialize(address_manager, self.peers_file_path)
+        except Exception:
+            self.log.exception("Error migrating legacy peer database")
```
I see, I didn't know that

arvidn commented:
Also, I think we should be careful to not aim *too* high in this patch. It sounds like we have a decent speed-up, and we should probably and this before trying to squeeze more performance out of this. It might not be the lowest hanging fruit anymore.

paninaro committed:
Minor tweaks to checking for migration

paninaro committed:
Removed AddressManagerSQLiteStore class scaffolding

paninaro commented on chia/server/address_manager_sqlite_store.py:
```diff
@@ -0,0 +1,119 @@
+        return [(node_id, bucket) for node_id, bucket in entries]
+
+    @classmethod
+    async def _is_empty(cls, metadata: Dict[str, str]) -> bool:
```
I'll change it. I do generally prefer to group related functions into a class (even if it's stateless), especially if I want to provide some semblance of a public/private interface, but if it's not the python way, that's fine.

arvidn replied:
I should have added that I'm not an authority on the "python way". I would say it would be a code smell in C++, but maybe it isn't in python.

paninaro committed:
makePeerDataSerialization now returns bytes instead of a PeerDataSerialization object

paninaro committed:
Async file I/O for write_file_async using aiofiles
Added more tests

paninaro committed:
Separate out the synchronous part of move_file

paninaro committed:
Renamed write_file to files since we're opening up the capabilities a bit

paninaro committed:
Update references to write_file

paninaro committed:
Renamed test_write_file to test_files

paninaro committed:
Tests covering move_file and move_file_async

paninaro committed:
Minor refinements to behavior and tests

paninaro committed:
Use aiofiles for reading peers.dat

paninaro commented:
> Also, I think we should be careful to not aim _too_ high in this patch. It sounds like we have a decent speed-up, and we should probably and this before trying to squeeze more performance out of this. It might not be the lowest hanging fruit anymore.
> 
> (this is a counter-weight against my comments about making the disk I/O async)

This was my original goal -- to address *just* the serialization format while squeezing some additional performance out at the same time. The data model could likely still use some tweaking to get the footprint down further. As it stands, the streamable-based serialization format is faster at serializing and nominally slower at deserializing (only occurring once during launch) when compared to sqlite. Serialization times were roughly 8x faster when I last checked.

arvidn commented:
the `aiofiles` dependency caused this:
```
chia/server/address_manager_store.py:1: error: Library stubs not installed for "aiofiles" (or incompatible with Python 3.8)
chia/server/address_manager_store.py:1: note: Hint: "python3 -m pip install types-aiofiles"
chia/server/address_manager_store.py:1: note: (or run "mypy --install-types" to install all missing stub packages)
```

paninaro committed:
Added missing mypy typing info for aiofiles. Also added types-PyYAML to dev_dependencies so that `mypy chia tests` doesn't require running with --install-types

paninaro committed:
Add types-aiofiles to the linting workflow

paninaro commented:
> the `aiofiles` dependency caused this:
> 
> ```
> chia/server/address_manager_store.py:1: error: Library stubs not installed for "aiofiles" (or incompatible with Python 3.8)
> chia/server/address_manager_store.py:1: note: Hint: "python3 -m pip install types-aiofiles"
> chia/server/address_manager_store.py:1: note: (or run "mypy --install-types" to install all missing stub packages)
> ```

Thanks for the pointer. Looks like a couple of the workflow files have their own sections for additional dependencies. I updated setup.py (dev_dependencies), upload-pypi-source.yml (which doesn't sound related, but...) and pre-commit-config.yaml.

arvidn commented:
it looks like the LGTM issues are legitimate ans should be addressed

paninaro commented:
> it looks like the LGTM issues are legitimate ans should be addressed

I find it strange that LGTM is complaining about these files now -- I haven't touched them in this PR. I'd rather not touch these files, given that they aren't related to the changes. Post-merge, I'm fine fixing the unused import, though the self-assignment issues identified in `timelord_state.py` should be looked at by someone knowledgable in that area.

arvidn commented:
I see

arvidn commented on chia/util/files.py:
```diff
@@ -0,0 +1,88 @@
+    temp_file_path: Path
+    async with tempfile.NamedTemporaryFile(dir=file_path.parent, mode=mode, delete=False) as f:
+        temp_file_path = f.name
+        await f.write(data)
```
I think it would be a good idea to add:
```
await f.flush()
```

here, to make sure the data actually hits the disk before swapping the files out, in case of catastrophic failures at the wrong time

paninaro replied:
I don't know that it would make a difference, but maybe. Since this is used in a context manager via `async with` the temp file will be closed before the files are swapped out and any buffers should be flushed in the process. Unless the close is a lie.

arvidn replied:
the [documentation](https://docs.python.org/3/library/io.html#io.IOBase.close) says:

> Flush and close this stream

And the documentation for `flush()` says:

> Flush the write buffers of the stream if applicable.

So, it's not explicit about whether it's also blocks until that flush completes, but it seems reasonable to assume that.

paninaro replied:
I'll add a call to flush to make it explicit.

arvidn commented on chia/util/files.py:
```diff
@@ -0,0 +1,88 @@
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(file_path.parent, mode=dir_perms, exist_ok=True)
```
is there a good reason to specify the permission bits here?
Do we specifically want to disallow the user from controlling this a more permissive `umask`?

paninaro replied:
Having a more secure default seems like the right behavior for the case that the intermediates don't already exist. For use cases that want to ensure something more open, they can create the intermediates before calling this. We could have the directory permissions passed into the function.

arvidn replied:
closing a file does not synchronize with the disk.

arvidn replied:
ehm. what I mean is `close(fd)` doesn't block until the data has hit the disk. it's not trivial to find authoritative documentation on closing files, but flushing would have to be an `async` function, and there's no `await` call corresponding to closing of the context manager, is there?

paninaro committed:
Directory perms can now be passed into write_file_async.
Added an explicit f.flush() followed by os.fsync() after writing the temp file contents.

wjblanke reviewed (approved):
aok
